### PR TITLE
Added `with_context` qualifier to `have_authorized_scope` matcher

### PR DESCRIPTION
**Pull request by killondark**
Hi @palkan. I added `with_context` qualifier to `have_authorized_scope` matcher
Discussed in https://github.com/palkan/action_policy/issues/258

PR checklist:

- [x] Tests included
- [x] Documentation updated

### CODE REVIEW

File: lib/action_policy/testing.rb [hunk 1]
```diff
@@ -34,21 +34,23 @@ def context_matches?(context, actual)
...
         end
 
-        def matches?(policy_class, actual_type, actual_name, actual_scope_options)
+        def matches?(policy_class, actual_type, actual_name, actual_scope_options, actual_context)
           policy_class == policy.class &&
             type == actual_type &&
             name == actual_name &&
-            actual_scope_options === scope_options
+            actual_scope_options === scope_options &&
+            actual_context.all? { |key, value| context[key] == value }
```
Comment (palkan) [inline lib/action_policy/testing.rb:53]: We should use the same logic here as for authorization calls: https://github.com/palkan/action_policy/blob/8e6923970a713bde164680df920522e98fff0d9b/lib/action_policy/testing.rb#L29-L33

It covers subsets and RSpec composable matchers.

File: lib/action_policy/rspec/have_authorized_scope.rb [hunk 2]
```diff
@@ -21,12 +21,13 @@ module RSpec
...
     class HaveAuthorizedScope < ::RSpec::Matchers::BuiltIn::BaseMatcher
       attr_reader :type, :name, :policy, :scope_options, :actual_scopes,
-        :target_expectations
+        :target_expectations, :context
 
       def initialize(type)
         @type = type
         @name = :default
         @scope_options = nil
+        @context = {}
```
Comment (palkan) [inline lib/action_policy/rspec/have_authorized_scope.rb:30]: The default MUST be `nil` (meaning  that we don't want to verify context)

File: lib/action_policy/testing.rb [hunk 3]
```diff
@@ -34,21 +34,23 @@ def context_matches?(context, actual)
...
-        attr_reader :policy, :target, :type, :name, :scope_options
+        attr_reader :policy, :target, :type, :name, :scope_options, :context
 
         def initialize(policy, target, type, name, scope_options)
           @policy = policy
           @target = target
           @type = type
           @name = name
           @scope_options = scope_options
+          @context = policy.authorization_context
```
Comment (palkan) [inline lib/action_policy/testing.rb:45]: We already have `policy` in the state, no need to extract context into its own instance variable.

File: docs/testing.md [hunk 4]
```diff
@@ -323,6 +323,10 @@ class UsersController < ApplicationController
   def index
     @user = authorized(User.all)
   end
+
+  def favorite
+    authorized_scope(User.all, context: {favorite: true})
+  end
```
Comment (palkan) [inline docs/testing.md:329]: A context is usually a record (it's an acting _subject_); it's unlikely a bool. Let's change this example:

```suggestion
  def for_user
    user = User.find(params[:id])
    authorized_scope(User.all, context: {user:})
  end
```

File: docs/testing.md [hunk 5]
```diff
@@ -408,6 +412,14 @@ expect { subject }.to have_authorized_scope(:scope)
   }
 ```
 
+Also can use the `with_context` options:
+
+```ruby
+expect { get :favorite }.to have_authorized_scope(:scope)
+  .with_scope_options(matching(with_deleted: a_falsey_value))
+  .with_context(favorite: true)
```
Comment (palkan) [inline docs/testing.md:420]:
```suggestion
expect { get :for_user, params: {id: user.id} }.to have_authorized_scope(:scope)
  .with_scope_options(matching(with_deleted: a_falsey_value))
  .with_context(a_hash_including(user:))
```

**Review by palkan — CHANGES_REQUESTED** [review]
Thanks!

Left some comments, please, take a look.

**Comment by killondark** [comment]
Thanks for your comments. Corrected. Please check it.

**Comment by killondark** [comment]
Hi @palkan. Have you seen my latest fixes?

**Review by palkan — APPROVED** [review]
Thanks!

**Pull request merged**